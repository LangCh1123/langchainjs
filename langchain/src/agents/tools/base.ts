import {CallbackManager, getCallbackManager} from "../../callbacks/index.js";

const getVerbosity = () => true;

export interface ToolParams {
  verbose?: boolean;
  callbackManager?: CallbackManager;
}

export abstract class Tool {

  verbose: boolean;

  callbackManager: CallbackManager;

  constructor(
      verbose?: boolean,
      callbackManager?: CallbackManager
  ) {
    this.verbose = verbose ?? getVerbosity();
    this.callbackManager = callbackManager ?? getCallbackManager();
  }

  protected abstract _call(arg: string): Promise<string>;

  async call(arg: string): Promise<string> {
    await this.callbackManager.handleToolStart({name: this.name}, arg, this.verbose);
    let result;
    try {
        result = await this._call(arg);
    } catch (e) {
        await this.callbackManager.handleToolError(e, this.verbose);
        throw e;
    }
    await this.callbackManager.handleToolEnd(result, this.verbose);
    return result;
  }

  abstract name: string;

  abstract description: string;

  returnDirect = false;
}
