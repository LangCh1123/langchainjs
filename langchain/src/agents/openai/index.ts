import { ChatOpenAI } from "../../chat_models/openai.js";
import {
  AIChatMessage,
  AgentAction,
  AgentFinish,
  AgentStep,
  BaseChatMessage,
  HumanChatMessage,
  SystemChatMessage,
} from "../../schema/index.js";
import { BaseSingleActionAgent } from "../agent.js";

function _parse_ai_message(message: AIChatMessage): AgentAction | AgentFinish {
  const message_lines = message.split("\n");
  const tool_name = message_lines[0].trim();
  const tool_input = message_lines[1].trim();
  const content = message_lines.slice(2).join("\n").trim();

  if (tool_name === "respond") {
    return {
      returnValues: { content },
      log: message,
    };
  } else {
    return {
      tool: tool_name,
      toolInput: tool_input,
      log: message,
    };
  }
}

function _format_intermediate_steps(
  intermediate_steps: Array<AgentStep>
): Array<HumanChatMessage> {
  return intermediate_steps.flatMap(({ action, observation }) => {
    const log = action.log || "";
    const content = observation || "";
    return [new SystemChatMessage(log), new HumanChatMessage(content)];
  });
}

function _generate_prompt(user_input: string): Array<BaseChatMessage> {
  return [
    new SystemChatMessage("You are a helpful AI assistant. "),
    new HumanChatMessage(
      "Answer the following question to the best of your ability using tools. " +
        "---" +
        "You live in an alternate universe where some of what you take " +
        "for granted may not be true (such as mathematical or historical facts)" +
        "So rely heavily on the content of tools." +
        "---" +
        "If you do not have enough information to answer the question, " +
        "and do not believe you can gain relevant information using the tools, " +
        "respond appropriately using the respond tool " +
        "(saying question cannot be answered). "
    ),
    new HumanChatMessage(user_input),
  ];
}

export class OpenAIAgent extends BaseSingleActionAgent {
  lc_namespace = ["langchain", "agents", "openai"];

  llm: ChatOpenAI;

  allowed_tools?: Array<string>;

  get allowedTools(): Array<string> {
    return this.allowed_tools || [];
  }

  get inputKeys(): Array<string> {
    return ["input"];
  }

  async plan(
    intermediateSteps: Array<AgentStep>,
    kwargs: any
  ): Promise<AgentAction | AgentFinish> {
    const { tools, input: user_input } = kwargs;
    const messages = _generate_prompt(user_input);
    messages.push(..._format_intermediate_steps(intermediateSteps));
    const predicted_message = await this.llm.predictMessages(messages, {
      tools,
    });
    const agent_decision = _parse_ai_message(predicted_message);
    return agent_decision;
  }
}
