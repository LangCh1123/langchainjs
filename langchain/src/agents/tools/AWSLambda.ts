import { Tool } from "./base.js";

interface LambdaConfig {
  region: string;
  accessKeyId: string;
  secretAccessKey: string;
  functionName: string;
}

class AWSLambda extends Tool {
  name: string;

  description: string;

  private lambdaConfig: LambdaConfig;

  constructor(
    toolName: string,
    toolDescription: string,
    lambdaConfig: LambdaConfig
  ) {
    super();

    this.name = toolName;
    this.description = toolDescription;
    this.lambdaConfig = lambdaConfig;
  }

  async _call(input: string): Promise<string> {
    const { Client, Invoker } = await LambdaImports();

    const lambdaClient = new Client({
      region: this.lambdaConfig.region,
      credentials: {
        accessKeyId: this.lambdaConfig.accessKeyId,
        secretAccessKey: this.lambdaConfig.secretAccessKey,
      },
    });

    return new Promise((resolve) => {
      const payloadUint8Array = new TextEncoder().encode(JSON.stringify(input));

      const command = new Invoker({
        FunctionName: this.lambdaConfig.functionName,
        InvocationType: "RequestResponse",
        Payload: payloadUint8Array,
      });

      lambdaClient
        .send(command)
        .then((response) => {
          const responseData = JSON.parse(
            new TextDecoder().decode(response.Payload)
          );

          resolve(responseData.body ? responseData.body : "request completed.");
        })
        .catch((error: Error) => {
          console.error("Error invoking Lambda function:", error);
          resolve("failed to complete request");
        });
    });
  }
}

async function LambdaImports() {
  try {
    const { LambdaClient, InvokeCommand } = await import(
      "@aws-sdk/client-lambda"
    );

    return {
      Client: LambdaClient as typeof LambdaClient,
      Invoker: InvokeCommand as typeof InvokeCommand,
    };
  } catch (e) {
    console.error(e);
    throw new Error(
      "Failed to load @aws-sdk/client-lambda'. Please install it eg. `yarn add @aws-sdk/client-lambda`."
    );
  }
}

export { AWSLambda };
