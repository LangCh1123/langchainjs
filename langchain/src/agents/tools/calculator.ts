import { Parser } from "expr-eval";

import { Tool } from "./base.js";

export class Calculator extends Tool {
  name = "calculator";

  async call(input: string) {
    try {
      return Parser.evaluate(input).toString();
    } catch (error: any) {
      return `javascript failed to parse. Try again: ${error.toString()}`;
    }
  }

  description = `Useful for getting the result of a math expression. 
  The input to this tool should be a well formed string math expression to be parsed and evaluated by expr-eval npm package."`;
}
