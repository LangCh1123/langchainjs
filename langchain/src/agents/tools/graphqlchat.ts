import { Tool } from "./base.js";

// Full documentation @ https://graphql.chat
const graphqlChatRestEndpoint = "https://graphql.chat/api";

class GraphqlChat extends Tool {
  name: string;

  description: string;

  accessToken?: string;

  graphqlEndpoint?: string;

  graphqlSchema?: string;

  graphqlEndpointAccessToken?: string;

  constructor({
    description,
    accessToken,
    graphqlEndpoint = process.env.GRAPHQLCHAT_ACCESS_TOKEN,
    graphqlSchema = process.env.GRAPHQLCHAT_SCHEMA,
    graphqlEndpointAccessToken = process.env.GRAPHQLCHAT_ENDPOINT_ACCESS_TOKEN,
  }: {
    description: string;
    accessToken?: string | undefined;
    graphqlEndpoint: string | undefined;
    graphqlSchema: string | undefined;
    graphqlEndpointAccessToken: string | undefined;
  }) {
    super();
    this.name = "GraphqlChat";
    this.description =
      description || "Query a remote graphql API to answer questions";
    this.accessToken = accessToken;
    this.graphqlEndpoint = graphqlEndpoint;
    this.graphqlSchema = graphqlSchema;
    this.graphqlEndpointAccessToken = graphqlEndpointAccessToken;
    if (!this.accessToken) {
      if (!this.graphqlEndpoint) {
        throw new Error(
          "graphqlEndpoint key not set. You can set it as GRAPHQLCHAT_ENDPOINT in your .env file."
        );
      }
      if (!this.graphqlSchema) {
        throw new Error(
          "graphqlSchema key not set. You can set it as GRAPHQLCHAT_SCHEMA in your .env file."
        );
      }
    }
  }

  async _call(input: string): Promise<string> {
    const response = await fetch(graphqlChatRestEndpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        // 'Authorization': `Bearer ${this.accessToken}` in case you got an access token from graphql.chat
        // Once you are using access token you can send only the question
      },
      body: JSON.stringify({
        question: input,
        graphqlEndpoint: this.graphqlEndpoint,
        graphqlSchema: this.graphqlSchema,
        endpointAccessToken: this.graphqlEndpointAccessToken,
      }),
    });
    const jsonResp = await response.json();
    if (jsonResp.errors) {
      throw new Error(`HTTP error ${jsonResp.errors[0].message}`);
    }
    return jsonResp.answer;
  }
}

export { GraphqlChat };
