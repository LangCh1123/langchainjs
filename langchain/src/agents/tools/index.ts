export { SerpAPI } from "./serpapi.js";
export { Calculator } from "./calculator.js";
export { DadJokeAPI } from "./dadjokeapi.js";
export { BingSerpAPI } from "./bingserpapi.js";
export { Tool } from "./base.js";
export { DynamicTool } from "./dynamic.js";
export { IFTTTWebhook } from "./IFTTTWebhook.js";
