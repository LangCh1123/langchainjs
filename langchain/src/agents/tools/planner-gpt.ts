import { AgentExecutor } from "agents/index.js";
import { DynamicTool } from "../../agents/tools/index.js";

// was hoping to make it a general recursive agent, but that tended to blow up, instead a planning agent to get it to trigger a act on a large plan from a small input
export const gptTool = (executor: AgentExecutor): DynamicTool => {
  const gpt = new DynamicTool({
    name: "gpt-agent",
    description:
      "useful to develop a strategic multistep plan to solve a larger task. Input is the outcome to acheive",
    func: async (input: string) => {
      console.log(`Executing with input "${input}"...`);

      const result = await executor.call({
        input: `Develop a strategic plan to ${input}. Consider the background information, constraints, and relevant context of the organization or industry. Provide specific requirements or preferences for the plan, such as timelines or budget constraints`,
        chat_history: [],
      });

      return Promise.resolve(result.output);
    },
  });

  return gpt;
};
