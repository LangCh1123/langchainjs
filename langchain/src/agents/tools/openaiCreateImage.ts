/** 
 * From https://platform.openai.com/docs/api-reference/images/create
 * 
 * To use this tool, you must pass in a configured OpenAIApi object.
 */
import { OpenAIApi } from "openai";

import { Tool } from "./base.js";

export class OpenAICreateImage extends Tool {
    constructor(readonly openaiApi: OpenAIApi, readonly name: string, readonly description: string) {
        super();
    }

    async call(input: string): Promise<string> {
        const resp = await this.openaiApi.createImage({
            prompt: input,
            // TODO: Future idea -- could we ask an LLM to extract these arguments from an input that might contain them?
            n: 1,
            size: "1024x1024",
        })

        const theImageUrl = resp.data.data[0].url;

        if (!theImageUrl) {
            throw new Error(`No image URL returned from OpenAI API.`);
        }

        return theImageUrl;
    }
}