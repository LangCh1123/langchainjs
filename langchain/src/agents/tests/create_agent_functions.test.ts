import { test, expect } from "@jest/globals";
import { ChatOpenAI, OpenAI } from "@langchain/openai";
import type {
  ChatPromptTemplate,
  PromptTemplate,
} from "@langchain/core/prompts";
import { TavilySearchResults } from "@langchain/community/tools/tavily_search";
import { pull } from "../../hub.js";
import {
  createOpenAIFunctionsAgent,
  createReactAgent,
  createStructuredChatAgent,
  createXmlAgent,
} from "../index.js";
import { ChatAnthropic } from "../../chat_models/anthropic.js";

const tools = [new TavilySearchResults({ maxResults: 1 })];

test("createStructuredChatAgent works", async () => {
  const prompt = await pull<ChatPromptTemplate>(
    "hwchase17/structured-chat-agent"
  );
  const llm = new ChatOpenAI({
    modelName: "gpt-3.5-turbo-1106",
    temperature: 0,
  });
  const agent = await createStructuredChatAgent({
    llm,
    tools,
    prompt,
  });
  const serializedAgent = agent.toJSON();
  expect(serializedAgent).toMatchSnapshot();
});

test("createOpenAIFunctionsAgent works", async () => {
  const prompt = await pull<ChatPromptTemplate>(
    "hwchase17/openai-functions-agent"
  );
  const llm = new ChatOpenAI({
    modelName: "gpt-3.5-turbo-1106",
    temperature: 0,
  });
  const agent = await createOpenAIFunctionsAgent({
    llm,
    tools,
    prompt,
  });
  const serializedAgent = agent.toJSON();
  expect(serializedAgent).toMatchSnapshot();
});

test("createOpenAIToolsAgent works", async () => {
  const prompt = await pull<ChatPromptTemplate>(
    "hwchase17/structured-chat-agent"
  );
  const llm = new ChatOpenAI({
    modelName: "gpt-3.5-turbo-1106",
    temperature: 0,
  });
  const agent = await createStructuredChatAgent({
    llm,
    tools,
    prompt,
  });
  const serializedAgent = agent.toJSON();
  expect(serializedAgent).toMatchSnapshot();
});

test("createXmlAgent works", async () => {
  const prompt = await pull<PromptTemplate>("hwchase17/xml-agent-convo");
  const llm = new ChatAnthropic({
    modelName: "claude-2.1",
    temperature: 0,
  });
  const agent = await createXmlAgent({
    llm,
    tools,
    prompt,
  });
  const serializedAgent = agent.toJSON();
  expect(serializedAgent).toMatchSnapshot();
});

test("createReactAgent works", async () => {
  const prompt = await pull<PromptTemplate>("hwchase17/react");
  const llm = new OpenAI({
    modelName: "gpt-3.5-turbo-instruct",
    temperature: 0,
  });
  const agent = await createReactAgent({
    llm,
    tools,
    prompt,
  });
  const serializedAgent = agent.toJSON();
  expect(serializedAgent).toMatchSnapshot();
});
