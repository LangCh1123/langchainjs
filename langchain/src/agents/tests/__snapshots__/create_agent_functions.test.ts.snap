// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createOpenAIFunctionsAgent works 1`] = `
{
  "id": [
    "langchain_core",
    "runnables",
    "RunnableSequence",
  ],
  "kwargs": {
    "first": {
      "id": [
        "langchain_core",
        "runnables",
        "RunnableAssign",
      ],
      "kwargs": {
        "mapper": {
          "id": [
            "langchain_core",
            "runnables",
            "RunnableMap",
          ],
          "kwargs": {
            "steps": {
              "agent_scratchpad": {
                "id": [
                  "langchain_core",
                  "runnables",
                  "RunnableLambda",
                ],
                "lc": 1,
                "type": "not_implemented",
              },
            },
          },
          "lc": 1,
          "type": "constructor",
        },
      },
      "lc": 1,
      "type": "constructor",
    },
    "last": {
      "id": [
        "langchain",
        "agents",
        "openai",
        "OpenAIFunctionsAgentOutputParser",
      ],
      "lc": 1,
      "type": "not_implemented",
    },
    "middle": [
      {
        "id": [
          "langchain_core",
          "prompts",
          "chat",
          "ChatPromptTemplate",
        ],
        "kwargs": {
          "input_variables": [
            "agent_scratchpad",
            "chat_history",
            "input",
          ],
          "messages": [
            {
              "id": [
                "langchain_core",
                "prompts",
                "chat",
                "SystemMessagePromptTemplate",
              ],
              "kwargs": {
                "prompt": {
                  "id": [
                    "langchain_core",
                    "prompts",
                    "prompt",
                    "PromptTemplate",
                  ],
                  "kwargs": {
                    "input_variables": [],
                    "partial_variables": undefined,
                    "template": "You are a helpful assistant",
                    "template_format": "f-string",
                  },
                  "lc": 1,
                  "type": "constructor",
                },
              },
              "lc": 1,
              "type": "constructor",
            },
            {
              "id": [
                "langchain_core",
                "prompts",
                "chat",
                "MessagesPlaceholder",
              ],
              "kwargs": {
                "optional": true,
                "variable_name": "chat_history",
              },
              "lc": 1,
              "type": "constructor",
            },
            {
              "id": [
                "langchain_core",
                "prompts",
                "chat",
                "HumanMessagePromptTemplate",
              ],
              "kwargs": {
                "prompt": {
                  "id": [
                    "langchain_core",
                    "prompts",
                    "prompt",
                    "PromptTemplate",
                  ],
                  "kwargs": {
                    "input_variables": [
                      "input",
                    ],
                    "partial_variables": undefined,
                    "template": "{input}",
                    "template_format": "f-string",
                  },
                  "lc": 1,
                  "type": "constructor",
                },
              },
              "lc": 1,
              "type": "constructor",
            },
            {
              "id": [
                "langchain_core",
                "prompts",
                "chat",
                "MessagesPlaceholder",
              ],
              "kwargs": {
                "optional": false,
                "variable_name": "agent_scratchpad",
              },
              "lc": 1,
              "type": "constructor",
            },
          ],
          "partial_variables": undefined,
        },
        "lc": 1,
        "type": "constructor",
      },
      {
        "id": [
          "langchain_core",
          "runnables",
          "RunnableBinding",
        ],
        "kwargs": {
          "bound": {
            "id": [
              "langchain",
              "chat_models",
              "openai",
              "ChatOpenAI",
            ],
            "kwargs": {
              "callbacks": undefined,
              "model": "gpt-3.5-turbo-1106",
              "openai_api_key": {
                "id": [
                  "OPENAI_API_KEY",
                ],
                "lc": 1,
                "type": "secret",
              },
              "temperature": 0,
              "verbose": undefined,
            },
            "lc": 1,
            "type": "constructor",
          },
          "config": {},
          "kwargs": {
            "functions": [
              {
                "description": "Useful for getting the result of a math expression. The input to this tool should be a valid mathematical expression that could be executed by a simple calculator.",
                "name": "calculator",
                "parameters": {
                  "$schema": "http://json-schema.org/draft-07/schema#",
                  "additionalProperties": false,
                  "properties": {
                    "input": {
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            ],
          },
        },
        "lc": 1,
        "type": "constructor",
      },
    ],
    "name": undefined,
  },
  "lc": 1,
  "type": "constructor",
}
`;

exports[`createOpenAIToolsAgent works 1`] = `
{
  "id": [
    "langchain_core",
    "runnables",
    "RunnableSequence",
  ],
  "kwargs": {
    "first": {
      "id": [
        "langchain_core",
        "runnables",
        "RunnableAssign",
      ],
      "kwargs": {
        "mapper": {
          "id": [
            "langchain_core",
            "runnables",
            "RunnableMap",
          ],
          "kwargs": {
            "steps": {
              "agent_scratchpad": {
                "id": [
                  "langchain_core",
                  "runnables",
                  "RunnableLambda",
                ],
                "lc": 1,
                "type": "not_implemented",
              },
            },
          },
          "lc": 1,
          "type": "constructor",
        },
      },
      "lc": 1,
      "type": "constructor",
    },
    "last": {
      "id": [
        "langchain",
        "agents",
        "structured_chat",
        "StructuredChatOutputParserWithRetries",
      ],
      "lc": 1,
      "type": "not_implemented",
    },
    "middle": [
      {
        "id": [
          "langchain_core",
          "prompts",
          "chat",
          "ChatPromptTemplate",
        ],
        "kwargs": {
          "input_variables": [
            "agent_scratchpad",
            "chat_history",
            "input",
          ],
          "messages": [
            {
              "id": [
                "langchain_core",
                "prompts",
                "chat",
                "SystemMessagePromptTemplate",
              ],
              "kwargs": {
                "prompt": {
                  "id": [
                    "langchain_core",
                    "prompts",
                    "prompt",
                    "PromptTemplate",
                  ],
                  "kwargs": {
                    "input_variables": [
                      "tool_names",
                      "tools",
                    ],
                    "partial_variables": undefined,
                    "template": "Respond to the human as helpfully and accurately as possible. You have access to the following tools:

{tools}

Use a json blob to specify a tool by providing an action key (tool name) and an action_input key (tool input).

Valid "action" values: "Final Answer" or {tool_names}

Provide only ONE action per $JSON_BLOB, as shown:

\`\`\`
{{
  "action": $TOOL_NAME,
  "action_input": $INPUT
}}
\`\`\`

Follow this format:

Question: input question to answer
Thought: consider previous and subsequent steps
Action:
\`\`\`
$JSON_BLOB
\`\`\`
Observation: action result
... (repeat Thought/Action/Observation N times)
Thought: I know what to respond
Action:
\`\`\`
{{
  "action": "Final Answer",
  "action_input": "Final response to human"
}}

Begin! Reminder to ALWAYS respond with a valid json blob of a single action. Use tools if necessary. Respond directly if appropriate. Format is Action:\`\`\`$JSON_BLOB\`\`\`then Observation",
                    "template_format": "f-string",
                  },
                  "lc": 1,
                  "type": "constructor",
                },
              },
              "lc": 1,
              "type": "constructor",
            },
            {
              "id": [
                "langchain_core",
                "prompts",
                "chat",
                "MessagesPlaceholder",
              ],
              "kwargs": {
                "optional": true,
                "variable_name": "chat_history",
              },
              "lc": 1,
              "type": "constructor",
            },
            {
              "id": [
                "langchain_core",
                "prompts",
                "chat",
                "HumanMessagePromptTemplate",
              ],
              "kwargs": {
                "prompt": {
                  "id": [
                    "langchain_core",
                    "prompts",
                    "prompt",
                    "PromptTemplate",
                  ],
                  "kwargs": {
                    "input_variables": [
                      "agent_scratchpad",
                      "input",
                    ],
                    "partial_variables": undefined,
                    "template": "{input}

{agent_scratchpad}
 (reminder to respond in a JSON blob no matter what)",
                    "template_format": "f-string",
                  },
                  "lc": 1,
                  "type": "constructor",
                },
              },
              "lc": 1,
              "type": "constructor",
            },
          ],
          "partial_variables": undefined,
        },
        "lc": 1,
        "type": "constructor",
      },
      {
        "id": [
          "langchain_core",
          "runnables",
          "RunnableBinding",
        ],
        "kwargs": {
          "bound": {
            "id": [
              "langchain",
              "chat_models",
              "openai",
              "ChatOpenAI",
            ],
            "kwargs": {
              "callbacks": undefined,
              "model": "gpt-3.5-turbo-1106",
              "openai_api_key": {
                "id": [
                  "OPENAI_API_KEY",
                ],
                "lc": 1,
                "type": "secret",
              },
              "temperature": 0,
              "verbose": undefined,
            },
            "lc": 1,
            "type": "constructor",
          },
          "config": {},
          "kwargs": {
            "stop": [
              "Observation",
            ],
          },
        },
        "lc": 1,
        "type": "constructor",
      },
    ],
    "name": undefined,
  },
  "lc": 1,
  "type": "constructor",
}
`;

exports[`createReactAgent works 1`] = `
{
  "id": [
    "langchain_core",
    "runnables",
    "RunnableSequence",
  ],
  "kwargs": {
    "first": {
      "id": [
        "langchain_core",
        "runnables",
        "RunnableAssign",
      ],
      "kwargs": {
        "mapper": {
          "id": [
            "langchain_core",
            "runnables",
            "RunnableMap",
          ],
          "kwargs": {
            "steps": {
              "agent_scratchpad": {
                "id": [
                  "langchain_core",
                  "runnables",
                  "RunnableLambda",
                ],
                "lc": 1,
                "type": "not_implemented",
              },
            },
          },
          "lc": 1,
          "type": "constructor",
        },
      },
      "lc": 1,
      "type": "constructor",
    },
    "last": {
      "id": [
        "langchain",
        "agents",
        "react",
        "ReActSingleInputOutputParser",
      ],
      "lc": 1,
      "type": "not_implemented",
    },
    "middle": [
      {
        "id": [
          "langchain_core",
          "prompts",
          "prompt",
          "PromptTemplate",
        ],
        "kwargs": {
          "input_variables": [
            "agent_scratchpad",
            "input",
          ],
          "partial_variables": undefined,
          "template": "Answer the following questions as best you can. You have access to the following tools:

{tools}

Use the following format:

Question: the input question you must answer
Thought: you should always think about what to do
Action: the action to take, should be one of [{tool_names}]
Action Input: the input to the action
Observation: the result of the action
... (this Thought/Action/Action Input/Observation can repeat N times)
Thought: I now know the final answer
Final Answer: the final answer to the original input question

Begin!

Question: {input}
Thought:{agent_scratchpad}",
          "template_format": "f-string",
        },
        "lc": 1,
        "type": "constructor",
      },
      {
        "id": [
          "langchain_core",
          "runnables",
          "RunnableBinding",
        ],
        "kwargs": {
          "bound": {
            "id": [
              "langchain",
              "llms",
              "openai",
              "OpenAI",
            ],
            "kwargs": {
              "callbacks": undefined,
              "model": "gpt-3.5-turbo-instruct",
              "openai_api_key": {
                "id": [
                  "OPENAI_API_KEY",
                ],
                "lc": 1,
                "type": "secret",
              },
              "temperature": 0,
              "verbose": undefined,
            },
            "lc": 1,
            "type": "constructor",
          },
          "config": {},
          "kwargs": {
            "stop": [
              "
Observation:",
            ],
          },
        },
        "lc": 1,
        "type": "constructor",
      },
    ],
    "name": undefined,
  },
  "lc": 1,
  "type": "constructor",
}
`;

exports[`createStructuredChatAgent works 1`] = `
{
  "id": [
    "langchain_core",
    "runnables",
    "RunnableSequence",
  ],
  "kwargs": {
    "first": {
      "id": [
        "langchain_core",
        "runnables",
        "RunnableAssign",
      ],
      "kwargs": {
        "mapper": {
          "id": [
            "langchain_core",
            "runnables",
            "RunnableMap",
          ],
          "kwargs": {
            "steps": {
              "agent_scratchpad": {
                "id": [
                  "langchain_core",
                  "runnables",
                  "RunnableLambda",
                ],
                "lc": 1,
                "type": "not_implemented",
              },
            },
          },
          "lc": 1,
          "type": "constructor",
        },
      },
      "lc": 1,
      "type": "constructor",
    },
    "last": {
      "id": [
        "langchain",
        "agents",
        "structured_chat",
        "StructuredChatOutputParserWithRetries",
      ],
      "lc": 1,
      "type": "not_implemented",
    },
    "middle": [
      {
        "id": [
          "langchain_core",
          "prompts",
          "chat",
          "ChatPromptTemplate",
        ],
        "kwargs": {
          "input_variables": [
            "agent_scratchpad",
            "chat_history",
            "input",
          ],
          "messages": [
            {
              "id": [
                "langchain_core",
                "prompts",
                "chat",
                "SystemMessagePromptTemplate",
              ],
              "kwargs": {
                "prompt": {
                  "id": [
                    "langchain_core",
                    "prompts",
                    "prompt",
                    "PromptTemplate",
                  ],
                  "kwargs": {
                    "input_variables": [
                      "tool_names",
                      "tools",
                    ],
                    "partial_variables": undefined,
                    "template": "Respond to the human as helpfully and accurately as possible. You have access to the following tools:

{tools}

Use a json blob to specify a tool by providing an action key (tool name) and an action_input key (tool input).

Valid "action" values: "Final Answer" or {tool_names}

Provide only ONE action per $JSON_BLOB, as shown:

\`\`\`
{{
  "action": $TOOL_NAME,
  "action_input": $INPUT
}}
\`\`\`

Follow this format:

Question: input question to answer
Thought: consider previous and subsequent steps
Action:
\`\`\`
$JSON_BLOB
\`\`\`
Observation: action result
... (repeat Thought/Action/Observation N times)
Thought: I know what to respond
Action:
\`\`\`
{{
  "action": "Final Answer",
  "action_input": "Final response to human"
}}

Begin! Reminder to ALWAYS respond with a valid json blob of a single action. Use tools if necessary. Respond directly if appropriate. Format is Action:\`\`\`$JSON_BLOB\`\`\`then Observation",
                    "template_format": "f-string",
                  },
                  "lc": 1,
                  "type": "constructor",
                },
              },
              "lc": 1,
              "type": "constructor",
            },
            {
              "id": [
                "langchain_core",
                "prompts",
                "chat",
                "MessagesPlaceholder",
              ],
              "kwargs": {
                "optional": true,
                "variable_name": "chat_history",
              },
              "lc": 1,
              "type": "constructor",
            },
            {
              "id": [
                "langchain_core",
                "prompts",
                "chat",
                "HumanMessagePromptTemplate",
              ],
              "kwargs": {
                "prompt": {
                  "id": [
                    "langchain_core",
                    "prompts",
                    "prompt",
                    "PromptTemplate",
                  ],
                  "kwargs": {
                    "input_variables": [
                      "agent_scratchpad",
                      "input",
                    ],
                    "partial_variables": undefined,
                    "template": "{input}

{agent_scratchpad}
 (reminder to respond in a JSON blob no matter what)",
                    "template_format": "f-string",
                  },
                  "lc": 1,
                  "type": "constructor",
                },
              },
              "lc": 1,
              "type": "constructor",
            },
          ],
          "partial_variables": undefined,
        },
        "lc": 1,
        "type": "constructor",
      },
      {
        "id": [
          "langchain_core",
          "runnables",
          "RunnableBinding",
        ],
        "kwargs": {
          "bound": {
            "id": [
              "langchain",
              "chat_models",
              "openai",
              "ChatOpenAI",
            ],
            "kwargs": {
              "callbacks": undefined,
              "model": "gpt-3.5-turbo-1106",
              "openai_api_key": {
                "id": [
                  "OPENAI_API_KEY",
                ],
                "lc": 1,
                "type": "secret",
              },
              "temperature": 0,
              "verbose": undefined,
            },
            "lc": 1,
            "type": "constructor",
          },
          "config": {},
          "kwargs": {
            "stop": [
              "Observation",
            ],
          },
        },
        "lc": 1,
        "type": "constructor",
      },
    ],
    "name": undefined,
  },
  "lc": 1,
  "type": "constructor",
}
`;

exports[`createXmlAgent works 1`] = `
{
  "id": [
    "langchain_core",
    "runnables",
    "RunnableSequence",
  ],
  "kwargs": {
    "first": {
      "id": [
        "langchain_core",
        "runnables",
        "RunnableAssign",
      ],
      "kwargs": {
        "mapper": {
          "id": [
            "langchain_core",
            "runnables",
            "RunnableMap",
          ],
          "kwargs": {
            "steps": {
              "agent_scratchpad": {
                "id": [
                  "langchain_core",
                  "runnables",
                  "RunnableLambda",
                ],
                "lc": 1,
                "type": "not_implemented",
              },
            },
          },
          "lc": 1,
          "type": "constructor",
        },
      },
      "lc": 1,
      "type": "constructor",
    },
    "last": {
      "id": [
        "langchain",
        "agents",
        "xml",
        "XMLAgentOutputParser",
      ],
      "lc": 1,
      "type": "not_implemented",
    },
    "middle": [
      {
        "id": [
          "langchain_core",
          "prompts",
          "chat",
          "ChatPromptTemplate",
        ],
        "kwargs": {
          "input_types": {},
          "input_variables": [
            "agent_scratchpad",
            "input",
          ],
          "messages": [
            {
              "id": [
                "langchain_core",
                "prompts",
                "chat",
                "HumanMessagePromptTemplate",
              ],
              "kwargs": {
                "prompt": {
                  "id": [
                    "langchain_core",
                    "prompts",
                    "prompt",
                    "PromptTemplate",
                  ],
                  "kwargs": {
                    "input_variables": [
                      "agent_scratchpad",
                      "chat_history",
                      "input",
                      "tools",
                    ],
                    "partial_variables": undefined,
                    "template": "You are a helpful assistant. Help the user answer any questions.

You have access to the following tools:

{tools}

In order to use a tool, you can use <tool></tool> and <tool_input></tool_input> tags. You will then get back a response in the form <observation></observation>
For example, if you have a tool called 'search' that could run a google search, in order to search for the weather in SF you would respond:

<tool>search</tool><tool_input>weather in SF</tool_input>
<observation>64 degrees</observation>

When you are done, respond with a final answer between <final_answer></final_answer>. For example:

<final_answer>The weather in SF is 64 degrees</final_answer>

Begin!

Previous Conversation:
{chat_history}

Question: {input}
{agent_scratchpad}",
                    "template_format": "f-string",
                  },
                  "lc": 1,
                  "type": "constructor",
                },
              },
              "lc": 1,
              "type": "constructor",
            },
          ],
          "output_parser": null,
          "partial_variables": undefined,
          "validate_template": false,
        },
        "lc": 1,
        "type": "constructor",
      },
      {
        "id": [
          "langchain_core",
          "runnables",
          "RunnableBinding",
        ],
        "kwargs": {
          "bound": {
            "id": [
              "langchain",
              "chat_models",
              "anthropic",
              "ChatAnthropic",
            ],
            "kwargs": {
              "anthropic_api_key": {
                "id": [
                  "ANTHROPIC_API_KEY",
                ],
                "lc": 1,
                "type": "secret",
              },
              "callbacks": undefined,
              "model": "claude-2.1",
              "temperature": 0,
              "verbose": undefined,
            },
            "lc": 1,
            "type": "constructor",
          },
          "config": {},
          "kwargs": {
            "stop": [
              "</tool_input>",
              "</final_answer>",
            ],
          },
        },
        "lc": 1,
        "type": "constructor",
      },
    ],
    "name": undefined,
  },
  "lc": 1,
  "type": "constructor",
}
`;
